Hangfire Notification Feature

Anthon123 opened the pull request:
Issue #3

Commit:
Add search filter functionality to dashboard.

* Will filter out specific failed jobs based on the arguments in each method call in each failed job.

* added automatic redirect to first page of FailedJobs when a new filter event occurs

* Added css styling to the filter textbox

* added filter on date functionality

* Divided Hangfire.js into 5 smaller files

* filter functionality implemented for enqueued, scheduled, processing, succeeded, failed and deleted job pages.

* removed redundant test project from solution

* Replaced overloaded Pager constructors with one constructor with default null values on parameters

* added state validation for JobCountByStateName() method

* javascript added for time picker

* Implemented text, date and time filtering on all pages. Works as long as no number is in the string!

* Sql bug fixed itself somehow. Fixed small bugs in code and changed visualisation style for dates on every job page.

* Feature switch working, set "EnableSearch" and "RelativeTime" when making the Dashboard options in Startup.cs

Commit:
new modules/classes for email notification

Commit:
another filter bug

Commit:
another filter bug

Commit:
Email notification on server timeout working. Set email and sendgrid settings in the web.config file within the <AppSettings> and <System.net> brackets.

Commit:
Email notification on server timeout working. Set email and sendgrid settings in the web.config file within the <AppSettings> and <System.net> brackets.

Commit:
added validation

Commit:
Added JobWatchdog background process - tbc

Commit:
Email notification on failed job peak should work. Set check interval and failed jobs threshold in the web.config file under <appSettings>, "failedJobCountThreshold"  and "failedJobCheckInterval"

Commit:
Email notification when a server is timed out or a server is experiencing a peak of failed jobs.
Configuration settings are set within the <appSettings>... </appSettings> brackets and the <mailSettings> ...</mailSettings> (see below) brackets in the web.config file.

Enable each of the functionalities by adding:
"<add key="enableFailedJobPeakNotification" value="true" />" and/or
"<add key="enableServerTimeOutNotification" value="true" />"
between <appSettings>... </appSettings>.

The configuration settings for email on failed job peaks are set by adding:
"<add key="failedJobCountThreshold" value="X" />" and
"<add key="failedJobCheckInterval" value="Y" />"
between <appSettings>... </appSettings>.
(X = maximum failed job count withinin the duration of Y minutes)

Email settings are set by adding:
"<add key="fromEmail" value="SomeEmail@testmail.com" />" and
"<add key="fromName" value="NameOfSender" />" and
"<add key="toEmail" value="receiver1@testemail.com,receiver2@testemail.com,..." />"
(NOTE:no spaces before or after the ',' character, fromEmail may be a fictional or real email adress)

and adding the following in the web.config file:
"<system.net>
    <mailSettings>
      <smtp from="test@domain.com">
        <network host="smtp.sendgrid.net" password="password" userName="username" port="587" />
      </smtp>
    </mailSettings>
</system.net>"
which is the sendgrid settings.

Commit:
removed test proj

stevetsang commented on src/Hangfire.Core/BackgroundJobServer.cs:
```diff
@@ -163,5 +174,47 @@ public void Start()
         {
         }
+
+        public void OnNext(int i)
```
what is i?

Anthon123 replied:
The interface didn't allow us to implement it without a type, "OnNext()" would had been preffered but did not work :/

stevetsang replied:
you can still name what **i** because still dunno what it is lol

RobertGill replied:
interesting one! maybe use string and default value that says "not used"?

stevetsang commented on src/Hangfire.Core/BackgroundJobServer.cs:
```diff
@@ -163,5 +174,47 @@ public void Start()
+
+        public void OnNext(int i)
+        {
+            int threshold = int.Parse(ConfigurationManager.AppSettings["failedJobCountThreshold"]);
```
use var please and also you should switch from int.Parse to int.TryParse there is a chance the data from appsettings could be wrong

stevetsang commented on src/Hangfire.Core/BackgroundJobServer.cs:
```diff
@@ -163,5 +174,47 @@ public void Start()
+            }
+            else if(_failedJobsCount > threshold )
+            {
+                string msg = "The server with machine name: \n";
```
use var and string interpolation.

stevetsang commented on src/Hangfire.Core/BackgroundJobServer.cs:
```diff
@@ -163,5 +174,47 @@ public void Start()
+        {
+            throw new NotImplementedException();
+        }
+
```
remove extra spaces here

stevetsang replied:
remove extra blank lines here.

stevetsang commented on src/Hangfire.Core/BackgroundJobServer.cs:
```diff
@@ -163,5 +174,47 @@ public void Start()
+
+        public void OnCompleted()
+        {
+            throw new NotImplementedException();
```
you sure that this will not just throw a random exception? also same with the one **above**.

stevetsang commented on src/Hangfire.Core/Server/ServerHeartbeat.cs:
```diff
@@ -26,6 +26,7 @@ internal class ServerHeartbeat : IBackgroundProcess
 
         public ServerHeartbeat(TimeSpan heartbeatInterval)
         {
+
```
extra white space?

stevetsang commented on src/Hangfire.Core/Server/ServerStatusNotifier.cs:
```diff
@@ -0,0 +1,54 @@
+
+            string[] receivers = toEmail.Split(',');
+
+            if (receivers != null || receivers.Count() > 0) {
```
{ on new line.

stevetsang commented on src/Hangfire.Core/Server/ServerWatchdog.cs:
```diff
@@ -16,13 +16,15 @@
     {
         public static readonly TimeSpan DefaultCheckInterval = TimeSpan.FromMinutes(5);
         public static readonly TimeSpan DefaultServerTimeout = TimeSpan.FromMinutes(5);
+        
```
remove extra line

Commit:
removed test proj...

stevetsang commented on src/Hangfire.Core/Server/ServerWatchdog.cs:
```diff
@@ -39,9 +41,24 @@ public void Execute(BackgroundProcessContext context)
                 {
+                    var value = ConfigurationManager.AppSettings["enableServerTimeOutNotification"];
+
+                    if (!string.IsNullOrEmpty(value) && Convert.ToBoolean(value) == true)
```
maybe use bool.tryparse here?

stevetsang commented on src/Hangfire.Core/Server/ServerWatchdog.cs:
```diff
@@ -39,9 +41,24 @@ public void Execute(BackgroundProcessContext context)
+
+                    if (!string.IsNullOrEmpty(value) && Convert.ToBoolean(value) == true)
+                    {
+                        string msg = "The servers with machine name and server ID: \n";
```
var and possibly use string interpolation too

stevetsang commented on src/Hangfire.SqlServer/SqlServerConnection.cs:
```diff
@@ -345,6 +345,20 @@ public override int RemoveTimedOutServers(TimeSpan timeOut)
+                throw new ArgumentException("The `timeOut` value must be positive.", "timeOut");
+            }
+
+            string query = string.Format(@"select Id from [{0}].Server where LastHeartbeat < @timeOutAt", _storage.GetSchemaName());
```
var

Commit:
code review changes.

Commit:
code review fixes

Commit:
fixed error

Commit:
renamed parameter

Commit:
Fix bugs from the search feature pull request

stevetsang commented on src/Hangfire.Core/BackgroundJobServer.cs:
```diff
@@ -163,5 +174,46 @@ public void Start()
+
+        public void OnCompleted()
+        {
+            throw new Exception("Method \"OnCompleted()\" not implemented.");
```
I don't mean put a text there I mean will this ever throw an exception because it will eventually complete and then break? logic doesn't seem right to me.

Anthon123 replied:
They will probably never be used so no. The exceptions are there in case anyone gets the idea to either OnError or OnCompleted while they are unimplemented.

stevetsang replied:
is there a better way for this? I have a feeling this functionality may not be pulled in due to this.

Commit:
bug after rebase

stevetsang commented on src/Hangfire.SqlServer/SqlServerConnection.cs:
```diff
@@ -498,5 +512,6 @@ public override List<string> GetAllItemsFromList(string key)
 
             return _storage.UseConnection(connection => connection.Query<string>(query, new { key = key }).ToList());
         }
+
```
remove extra space

Commit:
code review change

Commit:
replaced existing interface with custom one + install instructions below.

Configuration settings are set within the <appSettings>... </appSettings> brackets and the <mailSettings> ...</mailSettings> (see below) brackets in the web.config file.

Enable each of the functionalities by adding:
"<add key="enableFailedJobPeakNotification" value="true" />" and/or
"<add key="enableServerTimeOutNotification" value="true" />"
between <appSettings>... </appSettings>.

The configuration settings for email on failed job peaks are set by adding:
"<add key="failedJobCountThreshold" value="X" />" and
"<add key="failedJobCheckInterval" value="Y" />"
between <appSettings>... </appSettings>.
(X = maximum failed job count withinin the duration of Y minutes)

Email settings are set by adding:
"<add key="fromEmail" value="SomeEmail@testmail.com" />" and
"<add key="fromName" value="NameOfSender" />" and
"<add key="toEmail" value="receiver1@testemail.com,receiver2@testemail.com,..." />"
(NOTE:no spaces before or after the ',' character, fromEmail may be a fictional or real email adress)

and adding the following in the web.config file:
"<system.net>
    <mailSettings>
      <smtp from="test@domain.com">
        <network host="smtp.sendgrid.net" password="password" userName="username" port="587" />
      </smtp>
    </mailSettings>
</system.net>"
which is the sendgrid settings.

ghost commented:
Think we need to extract ServerStatusNotifier(interface out as INotifier to be used in Core) out of Core and into its own project for email notifications and then have an option in the core to be able to add different notification modules.  Also this will work well in the future when Hangfire support .net core as .net core does not support System.Net.Mail.

Commit:
separate email notification module halfway done

Commit:
modular email notification implemented, settings/featureswitch might need to be addded in BackgroundJobServerOptions

Commit:
gdfg

ghost commented:
closing pull request as this should be pulled into the new notificationmodule branch
